packages/common/src/utils/resolveDynamicValues.ts

```diff
@@ -0,0 +1,182 @@
+        break;
+      case "portal":
+        const portalValues = await resolvePortalValues(
+          valueDef as IDynamicPortalQueryDefinition,
+          context
+        );
+        result = { ...result, ...portalValues };
+        break;
+      case "service-query":
+        const serviceQueryValues = await resolveServiceQueryValues(
+          valueDef as IDynamicServiceQueryDefinition,
+          context
+        );
+        result = { ...result, ...serviceQueryValues };
+
+        break;
+      default:
+        return assertNever(valueDef, getProp(valueDef, "source"));
+    }
+  }
+
+  return result;
+}
+
+function assertNever(x: never, source: string): never {
+  throw new Error(`Unexpected source ${source}.`);
+}
+
+/**
+ * Resolve a set of dynamic values by extracting properties
+ * from the current user's portal/self
+ * @param valueDef
+ * @param context
+ * @returns
+ */
+async function resolvePortalValues(
+  valueDef: IDynamicPortalQueryDefinition,
+  context: IArcGISContext
+): Promise<DynamicValues> {
+  const result = {};
+  // TODO: This could use the context.portal values to skip even a single request
+  // get the portalSelf via memoized function
+  const portal = await memoizedPortalSelf(
+    context.requestOptions.authentication
+  );
+
+  // get the value from the portal and attach into the result
+  const value = getProp(portal, valueDef.sourcePath);
+  setProp(valueDef.outPath, value, result);
+
+  return result;
+}
+
+/**
+ * Resolve a set of dynamic values based on an item query.
+ * This is used to query for a set of items, extract a specific value from each
+ * then run an aggregation on the set of values.
+ * @param valueDef
+ * @param context
+ * @returns
+ */
+async function resolveItemQueryValues(
+  valueDef: IDynamicItemQueryDefinition,
+  context: IArcGISContext
+): Promise<DynamicValues> {
+  const result = {};
+  // const start = Date.now();
+  // execute the query
+  if (valueDef.options) {
+    const opts: ISearchOptions = {
+      ...(valueDef.options as ISearchOptions),
+      authentication: context.requestOptions.authentication,
+    };
+    const response = await memoizedItemSearch(opts);
+    const promises = await response.results.reduce(
+      async (valsPromise: any, r) => {
+        const vals = await valsPromise;
+        const v = getProp(r, valueDef.sourcePath);
+        if (typeof v === "string" || typeof v === "number") {
+          vals.push(Promise.resolve(v));
+        } else {
+          // v is itself a valueDefinition so we call resolveDynamicValues again
+          const vResult = await resolveDynamicValues([v], context);
+
+          // vals needs just the value - the number or string but vResult is an object
+          // so we need to get it out of the object
+          vals.push(getProp(vResult, v.outPath));
+        }
+
+        return vals;
+      },
+      Promise.resolve([] as any[])
+    );
+    const values = await Promise.all(promises);
```

Comment:
is any error handling needed for this and similar functions that may have unavailable items / layer queries, etc.

Reply:
see next comment - tldr; this is draft to get feedback / refactor before adding error trapping / tests etc